robottelo/config/validators.py

```diff
@@ -1,10 +1,75 @@
+            default=secrets.get('container_repo_quay_robot_token'),
+        ),
+        Validator('docker.private_registry_password', default=secrets.get('satqe_shared_password')),
+        Validator('ec2.access_key', default=secrets.get('ec2_access_key')),
+        Validator('ec2.secret_key', default=secrets.get('ec2_secret_key')),
+        Validator('gce.cert', default=secrets.get('gce_cert')),
+        Validator('git.password', default=secrets.get('gitea_user_password')),
+        Validator('git.github_token', default=secrets.get('github_read_token')),
+        Validator('http_proxy.password', default=secrets.get('satqe_shared_password')),
+        Validator('ipa.idm_password', default=secrets.get('idm_password')),
+        Validator('ipa.password', default=secrets.get('ipa_password')),
+        Validator('ipa.time_based_secret', default=secrets.get('totp_ipa_secret')),
+        Validator('ldap.password', default=secrets.get('ad_password')),
+        Validator('open_ldap.password', default=secrets.get('satqe_shared_password')),
+        Validator('osp.password', default=secrets.get('osp_password')),
+        Validator('osp.project_domain_id', default=secrets.get('osp_project_domain_id')),
+        Validator('report_portal.api_key', default=secrets.get('reportportal-robottelo-token')),
+        Validator('rh_cloud.token', default=secrets.get('rh-cloud-token')),
+        Validator('rhev.password', default=secrets.get('satqe_shared_password')),
+        Validator('rhev.ca_cert', default=secrets.get('rhev_cacert')),
+        Validator('rhev.image_password', default=secrets.get('satqe_shared_password')),
+        Validator('rhsso.rhsso_password', default=secrets.get('satqe_shared_password')),
+        Validator('rhsso.totp_secret', default=secrets.get('totp_rhsso_secret')),
+        Validator('remotedb.db_password', default=secrets.get('external_db_password')),
+        Validator(
+            'virtwho.esx.hypervisor_password', default=secrets.get('esx_hypervisor_password')
+        ),
+        Validator(
+            'virtwho.xen.hypervisor_password', default=secrets.get('xen_hypervisor_password')
+        ),
+        Validator(
+            'virtwho.hyperv.hypervisor_password', default=secrets.get('hyperv_hypervisor_password')
+        ),
+        Validator(
+            'virtwho.rhevm.hypervisor_password', default=secrets.get('rhevm_hypervisor_password')
+        ),
+        Validator(
+            'virtwho.libvirt.hypervisor_password',
+            default=secrets.get('libvirt_hypervisor_password'),
+        ),
+        Validator(
+            'virtwho.kubevirt.hypervisor_password',
+            default=secrets.get('kubevirt_hypervisor_password'),
+        ),
+        Validator('vmware.username', default=secrets.get('vmware_username')),
+        Validator('vmware.password', default=secrets.get('vmware_password')),
+        Validator('vmware.image_password', default=secrets.get('satqe_shared_password')),
```

Comment:
This is hardcoding all of the vault keys into source here, this is a huge downside to the approach of using validator defaults.

Reply:
Agree with hardcoding but I dont see how that could be a downside? 
The hardcoding in validators is the same as the hardcoded names of setting in validators!

Everyone in the team has access to secrets so shouldn't be an issue to access, and update the secrets.

Reply:
There is a big difference between the two, and its not about access/updating.

The names of the settings map directly to the yaml files and data structure within them. `vmware.image_password` as an argument for the validator is dictating both the structure of the configuration yaml and the acceptable values for a given field.

This structure that you've proposed now hardcodes the paths within vault, in our logic code, instead of within the config files themselves. Users of robottelo shouldn't have to modify this module to not use `satqe_shared_password` as their vault key, and they shouldn't be forced to use a key like `satqe_shared_password`.

Assuming I want to change the credential for this `vmware.image_password` field. It is now hardcoded in source as `satqe_shared_password`, and I can't update vault without impacting every other field mapped to this vault key, and/or updating source code.

This is violating the 12 factor app principles, which is a big reason for why we adopted dynaconf. Config should be cleanly and clearly separated from code, with the only 'overlap' being these validator paths and conf file location, because the config can't be completely arbitrary. 

@JacobCallahan and I discussed the option of trying to use dynaconf dynamic variable formatting for this, using `ssh_password: '@format {this.secrets.ssh_password}'` within the yaml file, and letting the dynaconf instance for `settings` object resolve from its own secrets loading, instead of creating two dynaconf instances. 

This isn't my decision alone to make, and I've thoroughly voiced my issues with this approach and have suggested the approach of using a yamlLoader to proxy these lookups like IQE has done.  If you and @JacobCallahan have heard these warnings and would like to proceed with this implementation, I won't stop you. But I think its setting us up for problems as-is in this PR.

Reply:
@mshriver RE:

> This structure that you've proposed now hardcodes the paths within vault, in our logic code, instead of within the config files themselves. Users of robottelo shouldn't have to modify this module to not use satqe_shared_password as their vault key, and they shouldn't be forced to use a key like satqe_shared_password.

Users can still use config files to not use secrets from vault, by just updating the setting in conf. infact,  version controlling the vault secret keys with validators will allow us to monitor the secret names changes in vault PnT store and hence we will have better control over modifying secrets from Sat rover group in the vault.

> Assuming I want to change the credential for this vmware.image_password field. It is now hardcoded in source as satqe_shared_password, and I can't update vault without impacting every other field mapped to this vault key, and/or updating source code.

Again, this comes under the version controlling the change in secrets and a better control as mentioned in my previous comment. Updating the vault for specific secret credentials and its impact on other field mapped to the key isnt an issue with the hardcoded approach itself. Its the issue of using same credentials for multiple secrets and thats issue even now with OCP secrets. To resolve it, we can have 1:1 mapping with secret and its credential even though the credentials are the same for multiple secrets.

> This is violating the 12 factor app principles, which is a big reason for why we adopted dynaconf. Config should be cleanly and clearly separated from code, with the only 'overlap' being these validator paths and conf file location, because the config can't be completely arbitrary.

TBH, I am not a fan of principles unless the refactor / approach makes things easier to access, monitor and control the credentials from Vault. 

> @JacobCallahan and I discussed the option of trying to use dynaconf dynamic variable formatting for this, using ssh_password: '@format {this.secrets.ssh_password}' within the yaml file, 

What I dont like the YAML approach is the user has to remember the format and the entire path e.g `this.secrets.ssh_password` in case the user updated the conf locally with some credential string and now wants to revert back so that the user can access secret from the vault. This way we are forcing users to remember the path or every time go and look into CI to pick copy the path to access secrets instead of automation(now Validators) doing it for us. 

> and letting the dynaconf instance for settings object resolve from its own secrets loading, instead of creating two dynaconf instances.

This could be the only reason I would go with the YAML approach but anyways they are two different objects of dynaConf settings and shouldn't be interfering with each other.

Reply:
The changes were made as requested after @JacobCallahan convinced me on some security aspects of it!